File: core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/result/iceberg/IcebergDocument.scala
package edu.uci.ics.amber.core.storage.result.iceberg

import edu.uci.ics.amber.core.storage.IcebergCatalogInstance
import edu.uci.ics.amber.core.storage.model.{BufferedItemWriter, VirtualDocument}
import edu.uci.ics.amber.core.storage.util.StorageUtil.{withLock, withReadLock, withWriteLock}
import edu.uci.ics.amber.util.IcebergUtil
import org.apache.iceberg.{FileScanTask, Table}
import org.apache.iceberg.catalog.{Catalog, TableIdentifier}
import org.apache.iceberg.data.Record
import org.apache.iceberg.exceptions.NoSuchTableException
import org.apache.iceberg.types.{Conversions, Types}

import java.net.URI
import java.util.concurrent.locks.{ReentrantLock, ReentrantReadWriteLock}
import scala.jdk.CollectionConverters._
import java.nio.ByteBuffer
import java.time.{Instant, LocalDate, ZoneOffset}
import java.time.format.DateTimeFormatter
import scala.collection.mutable

/**
  * IcebergDocument is used to read and write a set of T as an Iceberg table.
  * It provides iterator-based read methods and supports multiple writers to write to the same table.
  *
  * The table must exist when constructing the document
  *
  * @param tableNamespace namespace of the table.
  * @param tableName name of the table.
  * @param tableSchema schema of the table.
  * @param serde function to serialize T into an Iceberg Record.
  * @param deserde function to deserialize an Iceberg Record into T.
  * @tparam T type of the data items stored in the Iceberg table.
  */
private[storage] class IcebergDocument[T >: Null <: AnyRef](
    val tableNamespace: String,
    val tableName: String,
    val tableSchema: org.apache.iceberg.Schema,
    val serde: (org.apache.iceberg.Schema, T) => Record,
    val deserde: (org.apache.iceberg.Schema, Record) => T
) extends VirtualDocument[T] {

  private val lock = new ReentrantReadWriteLock()

  @transient lazy val catalog: Catalog = IcebergCatalogInstance.getInstance()

  /**
    * Returns the URI of the table location.
    * @throws NoSuchTableException if the table does not exist.
    */
  override def getURI: URI = {
    val table = IcebergUtil
      .loadTableMetadata(catalog, tableNamespace, tableName)
      .getOrElse(
        throw new NoSuchTableException(f"table ${tableNamespace}.${tableName} doesn't exist")
      )
    URI.create(table.location())
  }

  /**
    * Deletes the table and clears its contents.
    */
  override def clear(): Unit =
    withWriteLock(lock) {
      val identifier = TableIdentifier.of(tableNamespace, tableName)
      if (catalog.tableExists(identifier)) {
        catalog.dropTable(identifier)
      }
    }

  /**
    * Get an iterator for reading records from the table.
    */
  override def get(): Iterator[T] = getUsingFileSequenceOrder(0, None)

  /**
    * Get records within a specified range [from, until).
    */
  override def getRange(from: Int, until: Int): Iterator[T] = {
    getUsingFileSequenceOrder(from, Some(until))
  }

  /**
    * Get records starting after a specified offset.
    */
  override def getAfter(offset: Int): Iterator[T] = {
    getUsingFileSequenceOrder(offset, None)
  }

  /**
    * Get the total count of records in the table.
    */
  override def getCount: Long = {
    val table = IcebergUtil
      .loadTableMetadata(catalog, tableNamespace, tableName)
      .getOrElse(
        return 0
      )
    table.newScan().planFiles().iterator().asScala.map(f => f.file().recordCount()).sum
  }

  /**
    * Creates a BufferedItemWriter for writing data to the table.
    * @param writerIdentifier The writer's ID. It should be unique within the same table, as each writer will use it as
    *                         the prefix of the files they append
    */
  override def writer(writerIdentifier: String): BufferedItemWriter[T] = {
    new IcebergTableWriter[T](
      writerIdentifier,
      catalog,
      tableNamespace,
      tableName,
      tableSchema,
      serde
    )
  }

  /**
    * Util iterator to get T in certain range
    * @param from start from which record inclusively, if 0 means start from the first
    * @param until end at which record exclusively, if None means read to the table's EOF
    */
  private def getUsingFileSequenceOrder(from: Int, until: Option[Int]): Iterator[T] =
    withReadLock(lock) {
      new Iterator[T] {
        private val iteLock = new ReentrantLock()
        // Load the table instance, initially the table instance may not exist
        private var table: Option[Table] = loadTableMetadata()

        // Last seen snapshot id(logically it's like a version number). While reading, new snapshots may be created
        private var lastSnapshotId: Option[Long] = None

        // Counter for how many records have been skipped
        private var numOfSkippedRecords = 0

        // Counter for how many records have been returned
        private var numOfReturnedRecords = 0

        // Total number of records to return
        private val totalRecordsToReturn = until.map(_ - from).getOrElse(Int.MaxValue)

        // Iterator for usable file scan tasks
        private var usableFileIterator: Iterator[FileScanTask] = seekToUsableFile()

        // Current record iterator for the active file
        private var currentRecordIterator: Iterator[Record] = Iterator.empty

        // Util function to load the table's metadata
        private def loadTableMetadata(): Option[Table] = {
          IcebergUtil.loadTableMetadata(
            catalog,
            tableNamespace,
            tableName
          )
        }

        private def seekToUsableFile(): Iterator[FileScanTask] =
          withLock(iteLock) {
            if (numOfSkippedRecords > from) {
              throw new RuntimeException("seek operation should not be called")
            }

            // refresh the table's snapshots
            if (table.isEmpty) {
              table = loadTableMetadata()
            }
            table.foreach(_.refresh())

            // Retrieve and sort the file scan tasks by file sequence number
            val fileScanTasksIterator: Iterator[FileScanTask] = table match {
              case Some(t) =>
                val currentSnapshotId = Option(t.currentSnapshot()).map(_.snapshotId())
                val fileScanTasks = (lastSnapshotId, currentSnapshotId) match {
                  // Read from the start
                  case (None, Some(_)) =>
                    val tasks = t.newScan().planFiles().iterator().asScala
                    lastSnapshotId = currentSnapshotId
                    tasks

                  // Read incrementally from the last snapshot
                  case (Some(lastId), Some(currId)) if lastId != currId =>
                    val tasks = t
                      .newIncrementalAppendScan()
                      .fromSnapshotExclusive(lastId)
                      .toSnapshot(currId)
                      .planFiles()
                      .iterator()
                      .asScala
                    lastSnapshotId = currentSnapshotId
                    tasks

                  // No new data
                  case (Some(lastId), Some(currId)) if lastId == currId =>
                    Iterator.empty

                  // Default: No data yet
                  case _ =>
                    Iterator.empty
                }
                fileScanTasks.toSeq.sortBy(_.file().fileSequenceNumber()).iterator

              case None =>
                Iterator.empty
            }

            // Iterate through sorted FileScanTasks and update numOfSkippedRecords
            val usableTasks = fileScanTasksIterator.dropWhile { task =>
              val recordCount = task.file().recordCount()
              if (numOfSkippedRecords + recordCount <= from) {
                numOfSkippedRecords += recordCount.toInt
                true
              } else {
                false
              }
            }

            usableTasks
          }

        override def hasNext: Boolean = {
          if (numOfReturnedRecords >= totalRecordsToReturn) {
            return false
          }

          if (!usableFileIterator.hasNext) {
            usableFileIterator = seekToUsableFile()
          }

          while (!currentRecordIterator.hasNext && usableFileIterator.hasNext) {
            val nextFile = usableFileIterator.next()
            currentRecordIterator = IcebergUtil.readDataFileAsIterator(
              nextFile.file(),
              tableSchema,
              table.get
            )

            // Skip records within the file if necessary
            val recordsToSkipInFile = from - numOfSkippedRecords
            if (recordsToSkipInFile > 0) {
              currentRecordIterator = currentRecordIterator.drop(recordsToSkipInFile)
              numOfSkippedRecords += recordsToSkipInFile
            }
          }

          currentRecordIterator.hasNext
        }

        override def next(): T = {
          if (!hasNext) throw new NoSuchElementException("No more records available")

          val record = currentRecordIterator.next()
          numOfReturnedRecords += 1
          deserde(tableSchema, record)
        }
      }
    }

  /**
    * Provides methods for extracting metadata statistics for the results
    *
    * **Statistics Computed:**
    * - **Numeric fields (Int, Long, Double)**: Computes `min` and `max`.
    * - **Date fields (Timestamp)**: Computes `min` and `max` (converted to `LocalDate`).
    * - **All fields**: Computes `not_null_count` (number of non-null values).
    *
    * @return A map where each field name is mapped to a nested map containing its statistics.
    * @throws NoSuchTableException if the table does not exist in the catalog.
    */
  override def getTableStatistics: Map[String, Map[String, Any]] = {
    val table = IcebergUtil
      .loadTableMetadata(catalog, tableNamespace, tableName)
      .getOrElse(
        throw new NoSuchTableException(f"table ${tableNamespace}.${tableName} doesn't exist")
      )

    val schema = table.schema()

    // Extract field names, IDs, and types from the schema
    val fieldTypes =
      schema.columns().asScala.map(col => col.name() -> (col.fieldId(), col.`type`())).toMap
    val fieldStats = mutable.Map[String, mutable.Map[String, Any]]()
    val dateFormatter: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE

    // Initialize statistics for each field
    fieldTypes.foreach {
      case (field, (_, fieldType)) =>
        val initialStats = mutable.Map[String, Any](
          "not_null_count" -> 0L
        )
        if (
          fieldType == Types.IntegerType.get() || fieldType == Types.LongType
            .get() || fieldType == Types.DoubleType.get()
        ) {
          initialStats("min") = Double.MaxValue
          initialStats("max") = Double.MinValue
        } else if (
          fieldType == Types.TimestampType.withoutZone() || fieldType == Types.TimestampType
            .withZone()
        ) {
          initialStats("min") = LocalDate.MAX.format(dateFormatter)
          initialStats("max") = LocalDate.MIN.format(dateFormatter)
        }

        fieldStats(field) = initialStats
    }

    // Scan table files and aggregate statistics
    table.newScan().includeColumnStats().planFiles().iterator().asScala.foreach { file =>
      val fileStats = file.file()
      // Extract column-level statistics
      val lowerBounds =
        Option(fileStats.lowerBounds()).getOrElse(Map.empty[Integer, ByteBuffer].asJava)
      val upperBounds =
        Option(fileStats.upperBounds()).getOrElse(Map.empty[Integer, ByteBuffer].asJava)
      val nullCounts =
        Option(fileStats.nullValueCounts()).getOrElse(Map.empty[Integer, java.lang.Long].asJava)
      val nanCounts =
        Option(fileStats.nanValueCounts()).getOrElse(Map.empty[Integer, java.lang.Long].asJava)

      fieldTypes.foreach {
        case (field, (fieldId, fieldType)) =>
          val lowerBound = Option(lowerBounds.get(fieldId))
          val upperBound = Option(upperBounds.get(fieldId))
          val nullCount: Long = Option(nullCounts.get(fieldId)).map(_.toLong).getOrElse(0L)
          val nanCount: Long = Option(nanCounts.get(fieldId)).map(_.toLong).getOrElse(0L)
          val fieldStat = fieldStats(field)

          // Process min/max values for numerical types
          if (
            fieldType == Types.IntegerType.get() || fieldType == Types.LongType
              .get() || fieldType == Types.DoubleType.get()
          ) {
            lowerBound.foreach { buffer =>
              val minValue =
                Conversions.fromByteBuffer(fieldType, buffer).asInstanceOf[Number].doubleValue()
              fieldStat("min") = Math.min(fieldStat("min").asInstanceOf[Double], minValue)
            }

            upperBound.foreach { buffer =>
              val maxValue =
                Conversions.fromByteBuffer(fieldType, buffer).asInstanceOf[Number].doubleValue()
              fieldStat("max") = Math.max(fieldStat("max").asInstanceOf[Double], maxValue)
            }
          }
          // Process min/max values for timestamp types
          else if (
            fieldType == Types.TimestampType.withoutZone() || fieldType == Types.TimestampType
              .withZone()
          ) {
            lowerBound.foreach { buffer =>
              val epochMicros = Conversions
                .fromByteBuffer(Types.TimestampType.withoutZone(), buffer)
                .asInstanceOf[Long]
              val dateValue =
                Instant.ofEpochMilli(epochMicros / 1000).atZone(ZoneOffset.UTC).toLocalDate
              fieldStat("min") =
                if (
                  dateValue
                    .isBefore(LocalDate.parse(fieldStat("min").asInstanceOf[String], dateFormatter))
                )
                  dateValue.format(dateFormatter)
                else
                  fieldStat("min")
            }

            upperBound.foreach { buffer =>
              val epochMicros = Conversions
                .fromByteBuffer(Types.TimestampType.withoutZone(), buffer)
                .asInstanceOf[Long]
              val dateValue =
                Instant.ofEpochMilli(epochMicros / 1000).atZone(ZoneOffset.UTC).toLocalDate
              fieldStat("max") =
                if (
                  dateValue
                    .isAfter(LocalDate.parse(fieldStat("max").asInstanceOf[String], dateFormatter))
                )
                  dateValue.format(dateFormatter)
                else
                  fieldStat("max")
            }
          }
          // Update non-null count
          fieldStat("not_null_count") = fieldStat("not_null_count").asInstanceOf[Long] +
            (fileStats.recordCount().toLong - nullCount - nanCount)
      }
    }
    fieldStats.map {
      case (field, stats) =>
        field -> stats.toMap
    }.toMap
  }
}

File: core/amber/src/main/python/core/storage/iceberg/iceberg_document.py
from itertools import islice
from threading import RLock
from typing import Iterator, Optional, Callable, Iterable
from typing import TypeVar
from urllib.parse import ParseResult, urlparse

import pyarrow as pa
from pyiceberg.catalog import Catalog
from pyiceberg.schema import Schema
from pyiceberg.table import Table, FileScanTask
from readerwriterlock import rwlock

from core.storage.iceberg.iceberg_catalog_instance import IcebergCatalogInstance
from core.storage.iceberg.iceberg_table_writer import IcebergTableWriter
from core.storage.iceberg.iceberg_utils import (
    load_table_metadata,
    read_data_file_as_arrow_table,
)
from core.storage.model.virtual_document import VirtualDocument

# Define a type variable
T = TypeVar("T")


class IcebergDocument(VirtualDocument[T]):
    """
    IcebergDocument is used to read and write a set of T as an Iceberg table.
    It provides iterator-based read methods and supports multiple writers to write to
    the same table.

    - On construction, the table will be created if it does not exist.
    - If the table exists, it will be overridden.

    :param table_namespace: Namespace of the table.
    :param table_name: Name of the table.
    :param table_schema: Schema of the table.
    :param serde: A function to convert a T iterable into a pyarrow Table. Note the
    conversion is not based on a single T item (unlike Texera's Java IcebergDocument.)
    :param deserde: A function to convert a pyarrow Table back into a T iterable.
    """

    def __init__(
        self,
        table_namespace: str,
        table_name: str,
        table_schema: Schema,
        serde: Callable[[Schema, Iterable[T]], pa.Table],
        deserde: Callable[[Schema, pa.Table], Iterable[T]],
    ):
        self.table_namespace = table_namespace
        self.table_name = table_name
        self.table_schema = table_schema
        self.serde = serde
        self.deserde = deserde

        self.lock = rwlock.RWLockFair()
        self.catalog = IcebergCatalogInstance.get_instance()

    def get_uri(self) -> ParseResult:
        """Returns the URI of the table location."""
        table = load_table_metadata(self.catalog, self.table_namespace, self.table_name)
        if not table:
            raise Exception(
                f"table {self.table_namespace}.{self.table_name} doesn't exist."
            )
        return urlparse(table.location())

    def clear(self):
        """Deletes the table and clears its contents."""
        with self.lock.gen_wlock():
            table_identifier = f"{self.table_namespace}.{self.table_name}"
            if self.catalog.table_exists(table_identifier):
                self.catalog.drop_table(table_identifier)

    def get(self) -> Iterator[T]:
        """Get an iterator for reading all records from the table."""
        return self._get_using_file_sequence_order(0, None)

    def get_range(self, from_index: int, until_index: int) -> Iterator[T]:
        """Get records within a specified range [from, until)."""
        return self._get_using_file_sequence_order(from_index, until_index)

    def get_after(self, offset: int) -> Iterator[T]:
        """Get records starting after a specified offset."""
        return self._get_using_file_sequence_order(offset, None)

    def get_count(self) -> int:
        """Get the total count of records in the table."""
        table = load_table_metadata(self.catalog, self.table_namespace, self.table_name)
        if not table:
            return 0
        return sum(f.file.record_count for f in table.scan().plan_files())

    def writer(self, writer_identifier: str):
        """
        Creates a BufferedItemWriter for writing data to the table.
        :param writer_identifier: The writer's ID. It should be unique within the same
        table, as each writer will use it as the prefix of the files they append
        :return: An IcebergTableWriter
        """
        return IcebergTableWriter[T](
            writer_identifier=writer_identifier,
            catalog=self.catalog,
            table_namespace=self.table_namespace,
            table_name=self.table_name,
            table_schema=self.table_schema,
            serde=self.serde,
        )

    def _get_using_file_sequence_order(
        self, from_index: int, until_index: Optional[int]
    ) -> Iterator[T]:
        """Utility to get records within a specified range."""
        with self.lock.gen_rlock():
            return IcebergIterator[T](
                from_index,
                until_index,
                self.catalog,
                self.table_namespace,
                self.table_name,
                self.table_schema,
                self.deserde,
            )


class IcebergIterator(Iterator[T]):
    """
    A custom iterator class to read items from an iceberg table based on an index range.
    """

    def __init__(
        self,
        from_index: int,
        until_index: int,
        catalog: Catalog,
        table_namespace: str,
        table_name: str,
        table_schema: Schema,
        deserde: Callable[[Schema, pa.Table], Iterable[T]],
    ):
        self.from_index = from_index
        self.until_index = until_index
        self.catalog = catalog
        self.table_namespace = table_namespace
        self.table_name = table_name
        self.table_schema = table_schema
        self.deserde = deserde
        self.lock = RLock()
        # Counter for how many records have been skipped
        self.num_of_skipped_records = 0
        # Counter for how many records have been returned
        self.num_of_returned_records = 0
        # Total number of records to return, used for termination condition
        self.total_records_to_return = (
            self.until_index - self.from_index if until_index else float("inf")
        )
        # Load the table instance, initially the table instance may not exist
        self.table = self._load_table_metadata()
        # Iterator for usable file scan tasks
        self.usable_file_iterator = self._seek_to_usable_file()
        # Current record iterator for the active file
        self.current_record_iterator = iter([])

    def _load_table_metadata(self) -> Optional[Table]:
        """Util function to load the table's metadata."""
        return load_table_metadata(self.catalog, self.table_namespace, self.table_name)

    def _seek_to_usable_file(self) -> Iterator[FileScanTask]:
        """Find usable file scan tasks starting from the specified record index."""
        with self.lock:
            if self.num_of_skipped_records > self.from_index:
                raise RuntimeError("seek operation should not be called")

            # Load the table for the first time
            if not self.table:
                self.table = self._load_table_metadata()

            # If the table still does not exist after loading, end iterator.
            if self.table:
                try:
                    self.table.refresh()
                    current_snapshot = self.table.current_snapshot()
                    if current_snapshot is None:
                        return iter([])
                    sorted_file_scan_tasks = self._extract_sorted_file_scan_tasks(
                        current_snapshot
                    )
                    # Skip records in files before the `from_index`
                    for task in sorted_file_scan_tasks:
                        record_count = task.file.record_count
                        if (
                            self.num_of_skipped_records + record_count
                            <= self.from_index
                        ):
                            self.num_of_skipped_records += record_count
                            continue
                        yield task
                except Exception:
                    print("Could not read iceberg table:\n")
                    raise Exception
            else:
                return iter([])

    def _extract_sorted_file_scan_tasks(self, current_snapshot):
        """
        As self.table.inspect.entries() does not work with java files, this method
        implements the logic to find file_sequence_number for each data file ourselves
        :param current_snapshot: The current snapshot of the table.
        :return: The file scan tasks of the file sorted by file_sequence_number
        """
        file_sequence_map = {}
        for manifest in current_snapshot.manifests(self.table.io):
            for entry in manifest.fetch_manifest_entry(io=self.table.io):
                file_sequence_map[entry.data_file.file_path] = entry.sequence_number
        # Retrieve and sort the file scan tasks by file sequence number
        file_scan_tasks = list(self.table.scan().plan_files())
        # Sort files by their sequence number. Files without a sequence
        # number will be read last.
        sorted_file_scan_tasks = sorted(
            file_scan_tasks,
            key=lambda t: file_sequence_map.get(t.file.file_path, float("inf")),
        )
        return sorted_file_scan_tasks

    def __iter__(self) -> Iterator[T]:
        return self

    def __next__(self) -> T:
        if self.num_of_returned_records >= self.total_records_to_return:
            raise StopIteration("No more records available")

        while True:
            try:
                record = next(self.current_record_iterator)
                self.num_of_returned_records += 1
                return record
            except StopIteration:
                # current_record_iterator is exhausted, need to go to the next file
                try:
                    next_file = next(self.usable_file_iterator)
                    arrow_table = read_data_file_as_arrow_table(next_file, self.table)
                    self.current_record_iterator = self.deserde(
                        self.table_schema, arrow_table
                    )
                    # Skip records within the file if necessary
                    records_to_skip_in_file = (
                        self.from_index - self.num_of_skipped_records
                    )
                    if records_to_skip_in_file > 0:
                        self.current_record_iterator = self._skip_records(
                            self.current_record_iterator, records_to_skip_in_file
                        )
                        self.num_of_skipped_records += records_to_skip_in_file
                except StopIteration:
                    # no more files left in this table
                    raise StopIteration("No more records available")

    @staticmethod
    def _skip_records(iterator, count):
        return islice(iterator, count, None)
